src/meta/src/hummock/compaction/compaction_picker.rs

```diff
@@ -189,13 +190,131 @@ impl CompactionPicker for MinOverlappingPicker {
     }
 }
 
+pub struct ManualCompactionPicker {
+    compact_task_id: u64,
+    overlap_strategy: Arc<dyn OverlapStrategy>,
+    option: ManualCompactionOption,
+}
+
+impl ManualCompactionPicker {
+    pub fn new(
+        compact_task_id: u64,
+        overlap_strategy: Arc<dyn OverlapStrategy>,
+        option: ManualCompactionOption,
+    ) -> Self {
+        Self {
+            compact_task_id,
+            overlap_strategy,
+            option,
+        }
+    }
+}
+
+impl CompactionPicker for ManualCompactionPicker {
+    fn pick_compaction(
+        &self,
+        levels: &[Level],
+        level_handlers: &mut [LevelHandler],
+    ) -> Option<SearchResult> {
+        let level = self.option.level;
+        let target_level = level + 1;
+
+        let mut select_input_ssts = vec![];
+        let mut tmp_sst_info = SstableInfo::default();
+        let mut range_overlap_info = RangeOverlapInfo::default();
+        tmp_sst_info.key_range = Some(self.option.key_range.clone());
+        range_overlap_info.update(&tmp_sst_info);
+
+        let level_table_infos: Vec<SstableInfo> = levels[level]
+            .table_infos
+            .iter()
+            .filter(|sst_info| range_overlap_info.check_overlap(sst_info))
+            .filter(|sst_info| {
+                if self.option.internal_table_id.is_empty() {
+                    return true;
+                }
+
+                // to collect internal_table_id from sst_info
+                let table_id_in_sst: Vec<u32> = sst_info
+                    .vnode_bitmaps
+                    .iter()
+                    .map(|vmap| vmap.table_id)
+                    .collect();
+
+                // to filter sst_file by table_id
+                for table_id in &table_id_in_sst {
+                    if self.option.internal_table_id.contains(table_id) {
+                        return true;
+                    }
+                }
+
+                false
+            })
+            .cloned()
+            .collect();
+
+        for table in &level_table_infos {
+            if level_handlers[level].is_pending_compact(&table.id) {
+                continue;
+            }
+
+            let mut pending_campct = false;
+            let overlap_files = self
+                .overlap_strategy
+                .check_base_level_overlap(&[table.clone()], &levels[target_level].table_infos);
+            for other in overlap_files {
```

Comment:
you can rewrite this loop by `overlap_files.iter().any(|table| level_handlers[target_level].is_pending_compact(table.id))`